lib/livebook/apps/manager.ex

```diff
@@ -0,0 +1,245 @@
+
+    deployment = deployment_by_ref(state, ref)
+
+    case result do
+      {:ok, _pid} ->
+        {_, state} = pop_in(state.deployments[deployment.slug])
+        Logger.info("[app=#{deployment.slug}] Deployment successful, app running")
+        # Run sync in case there is already a new version for this app
+        {:noreply, sync_apps(state)}
+
+      {:error, error} ->
+        {:noreply, handle_deployment_failure(state, deployment, error)}
+    end
+  end
+
+  def handle_info({:retry, slug}, state) do
+    {:noreply, retry(state, slug)}
+  end
+
+  def handle_info(_message, state) do
+    {:noreply, state}
+  end
+
+  defp sync_apps(state) do
+    permanent_app_specs = Apps.get_permanent_app_specs()
+    state = deploy_missing_apps(state, permanent_app_specs)
+    close_leftover_apps(permanent_app_specs)
+    state
+  end
+
+  defp deploy_missing_apps(state, permanent_app_specs) do
+    for app_spec <- permanent_app_specs,
+        not Map.has_key?(state.deployments, app_spec.slug),
+        reduce: state do
+      state ->
+        case fetch_app(app_spec.slug) do
+          {:ok, _state, app} when app.app_spec.version == app_spec.version ->
+            state
+
+          {:ok, :reachable, app} ->
+            ref = redeploy(app, app_spec)
+            track_deployment(state, app_spec, ref)
+
+          {:ok, :unreachable, _app} ->
+            state
+
+          :error ->
+            ref = deploy(app_spec)
+            track_deployment(state, app_spec, ref)
+        end
+    end
+  end
+
+  defp close_leftover_apps(permanent_app_specs) do
+    permanent_slugs = MapSet.new(permanent_app_specs, & &1.slug)
+
+    for app <- Apps.list_apps(),
+        app.permanent,
+        app.slug not in permanent_slugs do
+      Livebook.App.close_async(app.pid)
+    end
+  end
+
+  defp fetch_app(slug) do
+    # We check both global and the tracker. The app may be present in
+    # the tracker for longer, but if it is actually down we will get
+    # the :app_closed event eventually. On the other hand, if the app
+    # is already somewhere, global will tell us sooner than tracker.
+    case Apps.fetch_app(slug) do
+      {:ok, app} ->
+        {:ok, :reachable, app}
+
+      :error ->
+        case Livebook.Tracker.fetch_app(slug) do
+          {:ok, app} -> {:ok, :unreachable, app}
+          :error -> :error
+        end
+    end
+  end
+
+  defp app_definitely_down?(slug) do
+    not Apps.exists?(slug) and Livebook.Tracker.fetch_app(slug) == :error
+  end
+
+  defp deploy(app_spec) do
+    deployer_pid = Livebook.Apps.Deployer.list_deployers() |> Enum.random()
+    Logger.info("[app=#{app_spec.slug}] Scheduling app deployment on node #{node(deployer_pid)}")
+
+    Livebook.Apps.Deployer.deploy_monitor(deployer_pid, app_spec,
+      permanent: true,
+      start_only: true
+    )
+  end
+
+  defp redeploy(app, app_spec) do
+    # Redeploying pushes the new notebook to an existing app process,
+    # so we need to run deployment on the same node
```

Comment:
When we redeploy, do we keep the slug on the same server? And if so, can we reuse the optimization for MIX PREVIOUS INSTALL DIR?